File: Sem3/Advanced Methods of Programming/MyInterpreter - CSharp Console Application/MyInterpreter CSharp/domain/statement/OpenStatement.cs
using System.IO;
using MyInterpreter_CSharp.domain.adt;

namespace MyInterpreter_CSharp.domain.statement
{
    public class OpenStatement : IStatement
    {
        private string _filename;
        private string _varFileId;
        private static int _uniqueKey = 1;

        public OpenStatement(string filename, string varFileId)
        {
            _filename = filename;
            _varFileId = varFileId;
        }

        public ProgramState Execute(ProgramState programState)
        {
            IMyDictionary<int, MyPair<string, TextReader>> fileTable = programState.FileTable;
            IMyDictionary<string, int> symbolTable = programState.SymbolTable;
            MyPair<string, TextReader> myPair;

            foreach (int key in fileTable.GetKeys())
            {
                myPair = fileTable.Get(key);
                if (myPair.First == _filename)
                    throw new StatementException("File already open.");
            }

            TextReader textReader = File.OpenText(_filename);
            myPair = new MyPair<string, TextReader>(_filename, textReader);
            fileTable.Add(_uniqueKey, myPair);
            symbolTable.Add(_varFileId, _uniqueKey);
            _uniqueKey++;

            return null;
        }

        public override string ToString()
        {
            return "openRFile(" + _varFileId + ",\"" + _filename + "\"";
        }
    }
}

File: Sem3/Advanced Methods of Programming/MyInterpreter - JavaFX - Final Project/project/src/domain/statement/FileStatements/OpenStatement.java
package domain.statement.FileStatements;


import domain.ProgramState;
import domain.adt.AdtException;
import domain.adt.MyDictionaryInterface;
import domain.adt.MyPair;
import domain.statement.StatementException;
import domain.statement.StatementInterface;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.util.Set;


public class OpenStatement implements StatementInterface
{
    private String filename;
    private String varFileId;
    private static int uniqueKey = 1;


    public OpenStatement(String filename, String varFileId)
    {
        this.filename = filename;
        this.varFileId = varFileId;
    }


    @Override
    public ProgramState execute(ProgramState programState) throws StatementException
    {
        MyDictionaryInterface<Integer, MyPair> fileTable = programState.getFileTable();
        MyDictionaryInterface<String, Integer> symbolTable = programState.getSymbolTable();
        Set<Integer> keys = fileTable.keySet();

        for (Integer key : keys)
        {
            try
            {
                MyPair pair = fileTable.get(key);
                if (filename.equals(pair.getFirst().toString()))
                    throw new StatementException(filename + " already exists in the file table.\nFile already in use.");
            }
            catch (AdtException ignored)
            {
            }
        }

        try
        {
            MyPair<String, BufferedReader> pair = new MyPair<>(filename, new BufferedReader(new FileReader(filename)));
            fileTable.add(uniqueKey, pair);
            symbolTable.add(varFileId, uniqueKey);
            uniqueKey++;
        }
        catch (FileNotFoundException exception)
        {
            throw new StatementException("Error while openning file. " + filename);
        }
        catch (AdtException exception)
        {
            throw new StatementException(exception.getMessage());
        }

        return null;
    }


    @Override
    public String toString()
    {
        return "OpenRFile(" + varFileId + ",\"" + filename + "\")";
    }
}
